src/Sarif/Visitors/VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT        
+// license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Sarif.VersionOne;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// This class visits all locations in a SARIF v1 log file. On observing use of result.logicalLocationKey,
+    /// which is used to disambiguate logical locations in the result.logicalLocations dictionary that have a 
+    /// common fully qualified name (but which are different types), the visitor creates a mapping between
+    /// the logical location key and its associated fully qualified name. This allows the v2 transformation
+    /// in particular to more easily populate its logical location equivalents. Additionally, the visitor 
+    /// stores a mapping from fully qualified name to decorated name, if one exists. These data (the
+    /// decorated name and fully qualified name) moved from the location object to the logical location
+    /// object in v2.
+    /// </summary>
+    public class VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor : SarifRewritingVisitorVersionOne
+    {
+        public VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor()
+        {
+            FullyQualifiedNameToDecoratedNameMap = new Dictionary<string, string>();
+            LogicalLocationKeyToFullyQualifiedNameMap = new Dictionary<string, string>();
+        }
+        public IDictionary<string, string> FullyQualifiedNameToDecoratedNameMap { get; set; }
+
+        public IDictionary<string, string> LogicalLocationKeyToFullyQualifiedNameMap { get; set; }
+
+        public override LocationVersionOne VisitLocationVersionOne(LocationVersionOne node)
+        {
+            if (!string.IsNullOrEmpty(node.LogicalLocationKey) &&
+                !string.IsNullOrEmpty(node.FullyQualifiedLogicalName) &&
+                !node.FullyQualifiedLogicalName.Equals(node.LogicalLocationKey))
+            {
+                LogicalLocationKeyToFullyQualifiedNameMap[node.LogicalLocationKey] = node.FullyQualifiedLogicalName;
+            }
+
+            if (!string.IsNullOrEmpty(node.DecoratedName))
+            {
+                string fullyQualifiedName = node.FullyQualifiedLogicalName ?? node.LogicalLocationKey;
```

Comment:
>LogicalLocationKey [](start = 83, length = 18)

Just use the key as the index, for reason given above.

Reply:
Sorry, that comment is wrong. But so it the code. :-) You're trying to figure out what to use as the key, so the variable name should be `key`, and the code is:
```
string key = node.LogicalLocationKey ?? node.FullyQualifiedLogicalName;
KeyToDecoratedNameMap[key] = node.DecoratedName;
```


---
In reply to: [244240704](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244240704) [](ancestors = 244240704)

Reply:
(This is the fix to the bug I was worried about.)

---
In reply to: [244373857](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244373857) [](ancestors = 244373857,244240704)

Reply:
Your statement isn't quite right, I mention the intent in the class comment. This code only works in cases where disambiguation is created, i.e., when a logical location key is used to resolve a collision. We don't need to persist all the other cases, because generally we have our hands on a fully qualified name already on those instances. Still open to making a change here if you see one.

---
In reply to: [244373857](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244373857) [](ancestors = 244373857,244240704)

Reply:
I still don't think this is right. You say that this code works "when a logical location key is used to resolve a collision." So consider the case:

```
location1 = {
    "fullyQualifiedLogicalName": "ambiguousFqln",
    "logicalLocationKey": "key1",
    "decoratedName": "dn1"
};

location2  = {
    "fullyQualifiedLogicalName": "ambiguousFqln",
    "logicalLocationKey": "key2",
    "decoratedName": "dn2"
};
```

`fullyQualifiedLogicalName` is the same, and non-null, for both locations. So when you visit `location1`, you will execute.
```
FullyQualifiedNameToDecoratedNameMap["ambiguousFqln"] = "dn1";
```
... and when you visit `location2`, you will execute
```
FullyQualifiedNameToDecoratedNameMap["ambiguousFqln"] = "dn2";
```
... losing `dn1` entirely. With my suggested code, you will execute
```
KeyNameToDecoratedNameMap["key1"] = "dn1";
...
KeyNameToDecoratedNameMap["key2"] = "dn2";
```

... which is what you need.

---
In reply to: [244377445](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244377445) [](ancestors = 244377445,244373857,244240704)

Reply:
I'd already fixed this. I was confused by your statement in stand-up, thought we were discussing the other mapping scenario (to retrieve FQN in cases where there is a key collision)

---
In reply to: [244383433](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244383433) [](ancestors = 244383433,244377445,244373857,244240704)